pkg/trimaran/lowriskovercommitment/lowriskovercommitment.go

```diff
@@ -0,0 +1,258 @@
+	// get all pods on node
+	nodeName := node.GetName()
+	podsOnNode := make([]*v1.Pod, 0)
+	pl.eventHandler.RLock()
+	for _, info := range pl.eventHandler.ScheduledPodsCache[node.GetName()] {
+		podsOnNode = append(podsOnNode, info.Pod)
+	}
+	pl.eventHandler.RUnlock()
+	// calculate risk based on requests and limits
+	nodeRequestsAndLimits := trimaran.GetNodeRequestsAndLimits(podsOnNode, node, pod)
+	riskCPU := pl.computeRisk(metrics, v1.ResourceCPU, watcher.CPU, node, nodeRequestsAndLimits)
+	riskMemory := pl.computeRisk(metrics, v1.ResourceMemory, watcher.Memory, node, nodeRequestsAndLimits)
+	rank := 1 - math.Max(riskCPU, riskMemory)
+
+	klog.V(6).InfoS("Node rank", "nodeName", nodeName, "riskCPU", riskCPU, "riskMemory", riskMemory, "rank", rank)
+
+	return rank
+}
+
+// computeRisk : calculate the risk of scheduling on node for a given resource
+func (pl *LowRiskOverCommitment) computeRisk(metrics []watcher.Metric, resourceName v1.ResourceName,
+	resourceType string, node *v1.Node, nodeRequestsAndLimits *trimaran.NodeRequestsAndLimits) float64 {
+
+	var riskLimit, riskLoad, totalRisk float64
+
+	defer func() {
+		klog.V(6).InfoS("Calculated risk", "node", klog.KObj(node), "resource", resourceName,
+			"riskLimit", riskLimit, "riskLoad", riskLoad, "totalRisk", totalRisk)
+	}()
+
+	nodeRequest := nodeRequestsAndLimits.NodeRequest
+	nodeLimit := nodeRequestsAndLimits.NodeLimit
+	nodeRequestMinusPod := nodeRequestsAndLimits.NodeRequestMinusPod
+	nodeLimitMinusPod := nodeRequestsAndLimits.NodeLimitMinusPod
+	nodeCapacity := nodeRequestsAndLimits.Nodecapacity
+
+	var request, limit, capacity, requestMinusPod, limitMinusPod int64
+	if resourceName == v1.ResourceCPU {
+		request = nodeRequest.MilliCPU
+		limit = nodeLimit.MilliCPU
+		requestMinusPod = nodeRequestMinusPod.MilliCPU
+		limitMinusPod = nodeLimitMinusPod.MilliCPU
+		capacity = nodeCapacity.MilliCPU
+	} else if resourceName == v1.ResourceMemory {
+		request = nodeRequest.Memory
+		limit = nodeLimit.Memory
+		requestMinusPod = nodeRequestMinusPod.Memory
+		limitMinusPod = nodeLimitMinusPod.Memory
+		capacity = nodeCapacity.Memory
+	} else {
+		// invalid resource
+		klog.V(6).InfoS("Unexpected resource", "resourceName", resourceName)
+		return 0
+	}
+
+	// (1) riskLimit : calculate overcommit potential load
+	excess := limit - request
+	if excess < 0 {
+		excess = 0
+	}
+	maxLoad := capacity
+	if limit < capacity {
+		maxLoad = limit
+	}
+	allowed := maxLoad - request
+	if allowed < 0 {
+		allowed = 0
+	}
+	if excess > 0 && allowed < excess {
+		riskLimit = 1 - float64(allowed)/float64(excess)
+	}
+	klog.V(6).InfoS("RiskLimit", "node", klog.KObj(node), "resource", resourceName, "riskLimit", riskLimit)
+
+	// (2) riskLoad : calculate measured overcommitment
+	zeroRequest := &framework.Resource{}
+	stats, ok := trimaran.CreateResourceStats(metrics, node, zeroRequest, resourceName, resourceType)
```

Comment:
Why pass in a zeroRequest? Does this mean the score of this portion won't take the incoming pod's limits into consideration?

Reply:
This part is about utilization. Since the pod hasn't started yet, we don't know how much utilization it will have. Currently, zero is assumed. We could use an estimate, but that would be a little risky.